File: packages/provide-uterm-go/deckmux/identity.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package deckmux

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"sort"
	"strconv"
	"strings"
)

// ResolvedIdentity mirrors provide.uterm.auth.ResolvedIdentity: an opaque,
// consumer-defined subject plus free-form claims and an SSH fingerprint.
type ResolvedIdentity struct {
	Subject     string
	Claims      map[string]any
	Fingerprint string
}

// supportedIdentityVersions is the frozenset of understood identity-frame
// protocol versions; unknown versions are ignored (forward-compat).
var supportedIdentityVersions = map[int]struct{}{1: {}}

// ParseIdentityFrame extracts a ResolvedIdentity from a control-channel frame,
// mirroring _identity.parse_identity_frame. It returns nil when the frame is
// not an "identity" message, the version is not understood, the subject is
// missing/empty/non-string, or (when expectedSecret is non-empty) the HMAC
// signature is missing or invalid. Malformed claims downgrade to an empty map
// rather than rejecting the identity.
func ParseIdentityFrame(frame map[string]any, expectedSecret []byte) *ResolvedIdentity {
	if s, _ := frame["type"].(string); s != "identity" {
		return nil
	}
	version, ok := identityVersion(frame["version"])
	if !ok {
		return nil
	}
	if _, supported := supportedIdentityVersions[version]; !supported {
		return nil
	}
	subject, ok := frame["subject"].(string)
	if !ok || subject == "" {
		return nil
	}
	claims := map[string]any{}
	if raw, ok := frame["claims"].(map[string]any); ok {
		for k, v := range raw {
			claims[k] = v
		}
	}
	fingerprint, _ := frame["fingerprint"].(string)

	if len(expectedSecret) > 0 {
		signature, ok := frame["signature"].(string)
		if !ok || signature == "" {
			return nil
		}
		transport, _ := frame["transport"].(string)
		claimsStr := pythonCompactJSON(claims)
		canonical := strconv.Itoa(version) + ":" + subject + ":" + fingerprint + ":" + transport + ":" + claimsStr
		mac := hmac.New(sha256.New, expectedSecret)
		mac.Write([]byte(canonical))
		expected := hex.EncodeToString(mac.Sum(nil))
		if !hmac.Equal([]byte(signature), []byte(expected)) {
			return nil
		}
	}

	return &ResolvedIdentity{Subject: subject, Claims: claims, Fingerprint: fingerprint}
}

// PresenceFromIdentity builds a UserPresence from a resolved identity,
// mirroring _identity.presence_from_identity. The user_id is always the raw
// subject; name/color/role fall back deterministically to connectionID-derived
// values when the claims don't supply them.
func PresenceFromIdentity(identity *ResolvedIdentity, connectionID string, takenColors map[string]struct{}, role string) UserPresence {
	claims := identity.Claims
	name := firstNonempty(
		strOrNone(claims["display_name"]),
		strOrNone(claims["display"]),
		nameFromSubject(identity.Subject),
	)
	if name == "" {
		name = GenerateName(connectionID)
	}
	color := strOrNone(claims["color"])
	if color == "" {
		color = GenerateColor(connectionID, takenColors)
	}
	resolvedRole := strOrNone(claims["role"])
	if resolvedRole == "" {
		resolvedRole = role
	}
	return UserPresence{
		UserID:   identity.Subject,
		Name:     name,
		Color:    color,
		Role:     resolvedRole,
		Initials: GenerateInitials(name),
	}
}

// IdentityPrincipal adapts a ResolvedIdentity to the duck-typed principal
// shape the DeckMuxPresence service consumes (SubjectID + DisplayName),
// mirroring _identity._IdentityPrincipal. It is returned by value; its
// identifying fields are unexported, giving Go value-copy immutability in
// place of Python's frozen dataclass.
type IdentityPrincipal struct {
	subjectID   string
	displayName string
	Identity    *ResolvedIdentity
}

// SubjectID returns the identity subject.
func (p IdentityPrincipal) SubjectID() string { return p.subjectID }

// DisplayName returns the resolved display name.
func (p IdentityPrincipal) DisplayName() string { return p.displayName }

// IdentityAsPrincipal adapts a ResolvedIdentity to an IdentityPrincipal,
// mirroring _identity.identity_as_principal.
func IdentityAsPrincipal(identity *ResolvedIdentity) IdentityPrincipal {
	claims := identity.Claims
	display := firstNonempty(
		strOrNone(claims["display_name"]),
		strOrNone(claims["display"]),
		nameFromSubject(identity.Subject),
	)
	if display == "" {
		display = identity.Subject
	}
	return IdentityPrincipal{
		subjectID:   identity.Subject,
		displayName: display,
		Identity:    identity,
	}
}

// ---------------------------------------------------------------------------
// Private helpers
// ---------------------------------------------------------------------------

// identityVersion coerces a frame's "version" value to an int, accepting the
// int and float64 forms JSON decoding may produce.
func identityVersion(v any) (int, bool) {
	switch n := v.(type) {
	case int:
		return n, true
	case int64:
		return int(n), true
	case float64:
		if n == float64(int(n)) {
			return int(n), true
		}
	}
	return 0, false
}

// firstNonempty returns the first non-empty string among values.
func firstNonempty(values ...string) string {
	for _, v := range values {
		if v != "" {
			return v
		}
	}
	return ""
}

// strOrNone coerces value to a non-empty stripped string, else "" (the Go
// stand-in for Python's None), mirroring _identity._str_or_none.
func strOrNone(value any) string {
	s, ok := value.(string)
	if !ok {
		return ""
	}
	return strings.TrimSpace(s)
}

// nameFromSubject extracts a display-friendly name from a subject like
// "sre:alice" → "alice", mirroring _identity._name_from_subject. A subject
// with no colon is used verbatim; an empty after-colon tail yields "".
func nameFromSubject(subject string) string {
	if !strings.Contains(subject, ":") {
		return strings.TrimSpace(subject)
	}
	_, tail, _ := strings.Cut(subject, ":")
	return strings.TrimSpace(tail)
}

// pythonCompactJSON serializes v the way Python's
// json.dumps(v, sort_keys=True, separators=(",", ":")) does (ensure_ascii,
// sorted object keys, no whitespace) so the HMAC canonical string matches the
// signer byte-for-byte.
func pythonCompactJSON(v any) string {
	var b strings.Builder
	encodePyJSON(&b, v)
	return b.String()
}

func encodePyJSON(b *strings.Builder, v any) {
	switch val := v.(type) {
	case nil:
		b.WriteString("null")
	case bool:
		if val {
			b.WriteString("true")
		} else {
			b.WriteString("false")
		}
	case string:
		encodePyString(b, val)
	case int:
		b.WriteString(strconv.Itoa(val))
	case int64:
		b.WriteString(strconv.FormatInt(val, 10))
	case float64:
		if val == float64(int64(val)) {
			b.WriteString(strconv.FormatInt(int64(val), 10))
		} else {
			b.WriteString(strconv.FormatFloat(val, 'g', -1, 64))
		}
	case []any:
		b.WriteByte('[')
		for i, e := range val {
			if i > 0 {
				b.WriteByte(',')
			}
			encodePyJSON(b, e)
		}
		b.WriteByte(']')
	case map[string]any:
		keys := make([]string, 0, len(val))
		for k := range val {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		b.WriteByte('{')
		for i, k := range keys {
			if i > 0 {
				b.WriteByte(',')
			}
			encodePyString(b, k)
			b.WriteByte(':')
			encodePyJSON(b, val[k])
		}
		b.WriteByte('}')
	default:
		b.WriteString("null")
	}
}

// encodePyString writes a JSON string escaped the way Python json.dumps does
// with ensure_ascii=True: short escapes for the common controls, \uXXXX for
// other control chars and all non-ASCII (surrogate pairs above U+FFFF).
func encodePyString(b *strings.Builder, s string) {
	b.WriteByte('"')
	for _, r := range s {
		switch r {
		case '"':
			b.WriteString(`\"`)
		case '\\':
			b.WriteString(`\\`)
		case '\n':
			b.WriteString(`\n`)
		case '\r':
			b.WriteString(`\r`)
		case '\t':
			b.WriteString(`\t`)
		case '\b':
			b.WriteString(`\b`)
		case '\f':
			b.WriteString(`\f`)
		default:
			switch {
			case r < 0x20:
				writeUnicodeEscape(b, r)
			case r < 0x7f:
				b.WriteRune(r)
			default:
				if r > 0xffff {
					r -= 0x10000
					writeUnicodeEscape(b, 0xd800+(r>>10))
					writeUnicodeEscape(b, 0xdc00+(r&0x3ff))
				} else {
					writeUnicodeEscape(b, r)
				}
			}
		}
	}
	b.WriteByte('"')
}

func writeUnicodeEscape(b *strings.Builder, r rune) {
	const hexDigits = "0123456789abcdef"
	b.WriteString(`\u`)
	b.WriteByte(hexDigits[(r>>12)&0xf])
	b.WriteByte(hexDigits[(r>>8)&0xf])
	b.WriteByte(hexDigits[(r>>4)&0xf])
	b.WriteByte(hexDigits[r&0xf])
}

File: packages/provide-uterm-csharp/src/Provide.Uterm/DeckMux/Identity.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace Provide.Uterm.DeckMux;

/// <summary>Opaque identity from a control-channel identity frame.</summary>
public sealed class ResolvedIdentity
{
    public required string Subject { get; init; }
    public Dictionary<string, object?> Claims { get; init; } = new();
    public string Fingerprint { get; init; } = "";
}

/// <summary>Collaborative presence record for a connected user.</summary>
public sealed class UserPresence
{
    public required string UserId { get; init; }
    public required string Name { get; init; }
    public required string Color { get; init; }
    public required string Role { get; init; }
    public required string Initials { get; init; }
}

/// <summary>Principal adapter for DeckMux presence services.</summary>
public sealed class IdentityPrincipal
{
    public string SubjectId { get; }
    public string DisplayName { get; }
    public ResolvedIdentity Identity { get; }

    public IdentityPrincipal(string subjectId, string displayName, ResolvedIdentity identity)
    {
        SubjectId = subjectId;
        DisplayName = displayName;
        Identity = identity;
    }
}

/// <summary>
/// Identity-frame parsing and presence derivation.
/// Port of packages/provide-uterm-go/deckmux/identity.go.
/// </summary>
public static class Identity
{
    private static readonly HashSet<int> SupportedIdentityVersions = [1];

    public static ResolvedIdentity? ParseIdentityFrame(
        IReadOnlyDictionary<string, object?> frame,
        byte[]? expectedSecret = null)
    {
        if (!frame.TryGetValue("type", out var typeObj) || typeObj as string != "identity")
        {
            return null;
        }

        if (!TryIdentityVersion(frame.GetValueOrDefault("version"), out var version))
        {
            return null;
        }

        if (!SupportedIdentityVersions.Contains(version))
        {
            return null;
        }

        if (frame.GetValueOrDefault("subject") is not string subject || subject.Length == 0)
        {
            return null;
        }

        var claims = new Dictionary<string, object?>();
        if (frame.GetValueOrDefault("claims") is IDictionary<string, object?> rawClaims)
        {
            foreach (var (k, v) in rawClaims)
            {
                claims[k] = v;
            }
        }
        else if (frame.GetValueOrDefault("claims") is Dictionary<string, object?> d)
        {
            foreach (var (k, v) in d)
            {
                claims[k] = v;
            }
        }

        var fingerprint = frame.GetValueOrDefault("fingerprint") as string ?? "";

        if (expectedSecret is { Length: > 0 })
        {
            if (frame.GetValueOrDefault("signature") is not string signature || signature.Length == 0)
            {
                return null;
            }

            var transport = frame.GetValueOrDefault("transport") as string ?? "";
            var claimsStr = PythonCompactJson(claims);
            var canonical = version.ToString(CultureInfo.InvariantCulture) + ":" + subject + ":" +
                            fingerprint + ":" + transport + ":" + claimsStr;
            using var hmac = new HMACSHA256(expectedSecret);
            var expected = Convert.ToHexString(hmac.ComputeHash(Encoding.UTF8.GetBytes(canonical)))
                .ToLowerInvariant();
            if (!CryptographicOperations.FixedTimeEquals(
                    Encoding.UTF8.GetBytes(signature),
                    Encoding.UTF8.GetBytes(expected)))
            {
                return null;
            }
        }

        return new ResolvedIdentity { Subject = subject, Claims = claims, Fingerprint = fingerprint };
    }

    public static UserPresence PresenceFromIdentity(
        ResolvedIdentity identity,
        string connectionId,
        ISet<string>? takenColors = null,
        string role = "viewer")
    {
        takenColors ??= new HashSet<string>();
        var claims = identity.Claims;
        var name = FirstNonempty(
            StrOrNone(claims.GetValueOrDefault("display_name")),
            StrOrNone(claims.GetValueOrDefault("display")),
            NameFromSubject(identity.Subject));
        if (name.Length == 0)
        {
            name = IdentityNames.GenerateName(connectionId);
        }

        var color = StrOrNone(claims.GetValueOrDefault("color"));
        if (color.Length == 0)
        {
            color = IdentityNames.GenerateColor(connectionId, takenColors);
        }

        var resolvedRole = StrOrNone(claims.GetValueOrDefault("role"));
        if (resolvedRole.Length == 0)
        {
            resolvedRole = role;
        }

        return new UserPresence
        {
            UserId = identity.Subject,
            Name = name,
            Color = color,
            Role = resolvedRole,
            Initials = IdentityNames.GenerateInitials(name),
        };
    }

    public static IdentityPrincipal IdentityAsPrincipal(ResolvedIdentity identity)
    {
        var claims = identity.Claims;
        var display = FirstNonempty(
            StrOrNone(claims.GetValueOrDefault("display_name")),
            StrOrNone(claims.GetValueOrDefault("display")),
            NameFromSubject(identity.Subject));
        if (display.Length == 0)
        {
            display = identity.Subject;
        }

        return new IdentityPrincipal(identity.Subject, display, identity);
    }

    private static bool TryIdentityVersion(object? v, out int version)
    {
        version = 0;
        switch (v)
        {
            case int i:
                version = i;
                return true;
            case long l:
                version = (int)l;
                return true;
            case double d when d == Math.Truncate(d):
                version = (int)d;
                return true;
            case JsonElement je when je.ValueKind == JsonValueKind.Number && je.TryGetInt32(out var n):
                version = n;
                return true;
            default:
                return false;
        }
    }

    private static string FirstNonempty(params string[] values)
    {
        foreach (var v in values)
        {
            if (v.Length > 0)
            {
                return v;
            }
        }

        return "";
    }

    private static string StrOrNone(object? value) =>
        value is string s ? s.Trim() : "";

    private static string NameFromSubject(string subject)
    {
        var idx = subject.IndexOf(':');
        if (idx < 0)
        {
            return subject.Trim();
        }

        return subject[(idx + 1)..].Trim();
    }

    internal static string PythonCompactJson(object? v)
    {
        var sb = new StringBuilder();
        EncodePyJson(sb, v);
        return sb.ToString();
    }

    private static void EncodePyJson(StringBuilder b, object? v)
    {
        switch (v)
        {
            case null:
                b.Append("null");
                break;
            case bool bl:
                b.Append(bl ? "true" : "false");
                break;
            case string s:
                EncodePyString(b, s);
                break;
            case int i:
                b.Append(i.ToString(CultureInfo.InvariantCulture));
                break;
            case long l:
                b.Append(l.ToString(CultureInfo.InvariantCulture));
                break;
            case double d when d == Math.Truncate(d):
                b.Append(((long)d).ToString(CultureInfo.InvariantCulture));
                break;
            case double d:
                b.Append(d.ToString("G", CultureInfo.InvariantCulture));
                break;
            case IList<object?> list:
                b.Append('[');
                for (var i = 0; i < list.Count; i++)
                {
                    if (i > 0)
                    {
                        b.Append(',');
                    }

                    EncodePyJson(b, list[i]);
                }

                b.Append(']');
                break;
            case IDictionary<string, object?> map:
            {
                var keys = map.Keys.OrderBy(k => k, StringComparer.Ordinal).ToList();
                b.Append('{');
                for (var i = 0; i < keys.Count; i++)
                {
                    if (i > 0)
                    {
                        b.Append(',');
                    }

                    EncodePyString(b, keys[i]);
                    b.Append(':');
                    EncodePyJson(b, map[keys[i]]);
                }

                b.Append('}');
                break;
            }
            default:
                b.Append("null");
                break;
        }
    }

    private static void EncodePyString(StringBuilder b, string s)
    {
        b.Append('"');
        foreach (var r in s)
        {
            switch (r)
            {
                case '"':
                    b.Append("\\\"");
                    break;
                case '\\':
                    b.Append("\\\\");
                    break;
                case '\n':
                    b.Append("\\n");
                    break;
                case '\r':
                    b.Append("\\r");
                    break;
                case '\t':
                    b.Append("\\t");
                    break;
                case '\b':
                    b.Append("\\b");
                    break;
                case '\f':
                    b.Append("\\f");
                    break;
                default:
                    if (r < 0x20)
                    {
                        WriteUnicodeEscape(b, r);
                    }
                    else if (r < 0x7f)
                    {
                        b.Append(r);
                    }
                    else if (r > 0xffff)
                    {
                        var v = r - 0x10000;
                        WriteUnicodeEscape(b, (char)(0xd800 + (v >> 10)));
                        WriteUnicodeEscape(b, (char)(0xdc00 + (v & 0x3ff)));
                    }
                    else
                    {
                        WriteUnicodeEscape(b, r);
                    }

                    break;
            }
        }

        b.Append('"');
    }

    private static void WriteUnicodeEscape(StringBuilder b, int r)
    {
        const string hex = "0123456789abcdef";
        b.Append("\\u");
        b.Append(hex[(r >> 12) & 0xf]);
        b.Append(hex[(r >> 8) & 0xf]);
        b.Append(hex[(r >> 4) & 0xf]);
        b.Append(hex[r & 0xf]);
    }
}
